File: app/src/main/java/de/project/lukas/model/Device.java
package de.project.lukas.model;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import java.util.Objects;

public abstract class Device {
    private final MutableLiveData<String> name = new MutableLiveData<>("");
    private final MutableLiveData<String> message = new MutableLiveData<>("");   // 2024-01-16
    private final MutableLiveData<Boolean> isConnected = new MutableLiveData<>(true);
    private final MutableLiveData<Integer> battery = new MutableLiveData<>(0);

    public LiveData<String> getName() {
        return name;
    }

    public LiveData<String> getMessage() {
        return message;
    }                   // 2024-01-17 Ralf

    public LiveData<Boolean> getStatus() {
        return isConnected;
    }

    public LiveData<Integer> getBattery() {
        return battery;
    }

    protected void setInitialName(String value) {
        name.setValue(value);
    }

    protected void setName(String value) {
        name.postValue(value);
    }

    protected void setMessage(String value) {
        message.postValue(value);
    }                             // 2024-01-17 Ralf

    protected void setBattery(int value) {
        battery.postValue(value);
    }

    protected void setIsConnected(boolean value) {
        isConnected.postValue(value);
    }

    public abstract String getAddress();

    public abstract void disconnect();

    public void switchOff() {};

    @NonNull
    @Override
    public String toString() {
        return Objects.requireNonNull(name.getValue());
    }
}

File: app/src/main/java/de/project/lukas/model/Device.kt
package de.project.lukas.model

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * Base type for every connected Bluetooth device. State is exposed as [StateFlow]
 * so the Compose UI can collect it lifecycle-aware, independent of any Android view.
 */
abstract class Device {
    private val _name = MutableStateFlow("")
    val name: StateFlow<String> = _name.asStateFlow()

    private val _message = MutableStateFlow("")
    val message: StateFlow<String> = _message.asStateFlow()

    private val _isConnected = MutableStateFlow(true)
    val isConnected: StateFlow<Boolean> = _isConnected.asStateFlow()

    private val _battery = MutableStateFlow(0)
    val battery: StateFlow<Int> = _battery.asStateFlow()

    abstract val address: String

    protected fun setName(value: String) {
        _name.value = value
    }

    protected fun setMessage(value: String) {
        _message.value = value
    }

    protected fun setBattery(value: Int) {
        _battery.value = value
    }

    protected fun setConnected(value: Boolean) {
        _isConnected.value = value
    }

    abstract fun disconnect()

    open fun switchOff() {}

    override fun toString(): String = _name.value
}
